lib/Sema/CSSimplify.cpp

```diff
@@ -474,91 +483,70 @@ matchCallArguments(SmallVectorImpl<AnyFunctionType::Param> &args,
       }
     }
 
-    // Okay, we've matched all the labeled closures; scan backwards from
-    // there to match the unlabeled trailing closure.
-    Optional<unsigned> unlabeledParamIdx;
-    if (prevParamIdx > 0) {
-      unsigned paramIdx = prevParamIdx - 1;
-
-      bool lastAcceptsTrailingClosure =
-          acceptsTrailingClosure(params[paramIdx]);
+    // Temporarily claim the unlabeled trailing closure so we won't bind to it
+    // while dealing with other params.
+    claimedArgs[unlabeledArgIdx] = true;
+  }
 
-      // If the last parameter is defaulted, this might be
-      // an attempt to use a trailing closure with previous
-      // parameter that accepts a function type e.g.
-      //
-      // func foo(_: () -> Int, _ x: Int = 0) {}
-      // foo { 42 }
-      //
-      // FIXME: shouldn't this skip multiple arguments and look for
-      // something that acceptsTrailingClosure rather than just something
-      // with a function type?
-      if (!lastAcceptsTrailingClosure && paramIdx > 0 &&
-          paramInfo.hasDefaultArgument(paramIdx)) {
-        auto paramType = params[paramIdx - 1].getPlainType();
-        // If the parameter before defaulted last accepts.
-        if (paramType->is<AnyFunctionType>()) {
-          lastAcceptsTrailingClosure = true;
-          paramIdx -= 1;
-        }
+  {
+    unsigned nextArgIdx = 0;
+    // Walk through the parameters, binding any unbound.
+    for (auto paramIdx : indices(params)) {
+      if (parameterBindings[paramIdx].empty()) {
+        bindNextParameter(paramIdx, nextArgIdx, false);
       }
+    }
+  }
 
-      if (lastAcceptsTrailingClosure)
-        unlabeledParamIdx = paramIdx;
-    }
-
-    // If there's no suitable last parameter to accept the trailing closure,
-    // notify the listener and bail if we need to.
-    if (!unlabeledParamIdx) {
-      // Try to use a specialized diagnostic for an extra closure.
-      bool isExtraClosure = false;
-      if (prevParamIdx == 0) {
-        isExtraClosure = true;
-      } else if (unlabeledArgIdx > 0) {
-        // Argument before the trailing closure.
-        unsigned prevArg = unlabeledArgIdx - 1;
-        auto &arg = args[prevArg];
-        // If the argument before trailing closure matches
-        // last parameter, this is just a special case of
-        // an extraneous argument.
-        const auto param = params[prevParamIdx - 1];
-        if (param.hasLabel() && param.getLabel() == arg.getLabel()) {
-          isExtraClosure = true;
-        }
-      }
+  // Okay, we've matched all other arguments; scan backwards from
+  // labeled closures to find a param for the unlabeled trailing closure.
+  if (unlabeledTrailingClosureArgIndex) {
+    unsigned unlabeledArgIdx = *unlabeledTrailingClosureArgIndex;
+    int paramIdx = prevParamIdx - 1;
```

Comment:
I think instead of `prevParamIdx` `paramIdx` should be `numParams - 1` because in multiple trailing closures case `prevParamIdx` would point to the leftmost labeled closure parameter. `prevParamIdx` should stay in scope of `if (unlabeledTrailingClosureArgIndex) {` like before to avoid confusion...

Reply:
At this point in the matching, any labeled trailing closures are matched, and we want to skip those params at the end that have already matched. The previous code here did the same thing when looking at the unlabeled closure (using `prevParamIdx - 1`) and I haven't changed that.